ocfweb/lab_reservations/reserve.py

```diff
@@ -0,0 +1,178 @@
+from collections import namedtuple
+
+import ocflib.misc.validators
+from django import forms
+from django.shortcuts import render
+from jinja2 import Environment
+from jinja2 import PackageLoader
+from ocflib.account.search import user_attrs_ucb
+from ocflib.misc.mail import send_mail
+
+from ocfweb.auth import calnet_required
+from ocfweb.component.forms import wrap_validator
+
+
+JINJA_MAIL_ENV = Environment(loader=PackageLoader('ocfweb', ''))
+
+
+class NewReservationRequest(namedtuple(
+    'NewReservationRequest', [
+        'real_name',
+        'contact_email',
+        'student_group',
+        'reason',
+        'date',
+        'starttime',
+        'endtime',
+    ],
+)):
+    """Request for reservation request.
+    :param real_name:
+    :param contact_email:
+    :param student_group:
+    :param reason:
+    :param date:
+    :param starttime:
+    :param endtime:
+    """
+
+    def to_dict(self):
+        return {
+            field: getattr(self, field)
+            for field in self._fields
+        }
+
+
+class RequestForm(forms.Form):
+
+    error_css_class = 'error'
+    required_css_class = 'required'
+
+    real_name = forms.CharField(
+        label='Full Name',
+        widget=forms.TextInput(attrs={'placeholder': 'Jane Doe'}),
+        min_length=3,
+        max_length=32,
+    )
+
+    contact_email = forms.EmailField(
+        label='Contact e-mail',
+        validators=[wrap_validator(ocflib.misc.validators.valid_email)],
+        widget=forms.EmailInput(attrs={'placeholder': 'janedoe@berkeley.edu'}),
+    )
+
+    verify_contact_email = forms.EmailField(
+        label='Confirm contact e-mail',
+        widget=forms.EmailInput(attrs={'placeholder': 'janedoe@berkeley.edu'}),
+    )
+
+    student_group = forms.CharField(
+        label='Student Group',
+        widget=forms.TextInput(attrs={'placeholder': 'OCF'}),
+        min_length=3,
+        max_length=32,
+    )
+
+    reason = forms.CharField(
+        label='Reason for reservation',
+        widget=forms.TextInput(attrs={'placeholder': ''}),
+    )
+
+    date = forms.DateField(
+        label='Date of reservation (mm/dd/yy)',
+        widget=forms.DateInput(attrs={'placeholder': '02/28/18'}),
+    )
+
+    starttime = forms.TimeField(
+        label='Starting time of reservation (xx:xx)',
+    )
+
+    endtime = forms.TimeField(
+        label='Ending time of reservation (xx:xx)',
+    )
```

Comment:
I assume these are 24-hour times? It's probably worth specifying. It's sort of obvious to people in technical fields because we use 24-hour clocks all the time, but it may be confusing to others ("why won't it let me enter 'PM'???")

Reply:
Err... I don't know if 24 hour format is our convention, but I'd say use 12 hour format -- that's what most people are used to. You can pass `input_formats` to a `TimeField` to change the formatting. [Here](http://blog.tkbe.org/archive/date-filter-cheat-sheet/) is a cheat sheet. `%g` and `%A` will serve your purpose.